### ISSUE TITLE
pm.Deterministic

### ISSUE THREAD

**Issue by erlebach** [pymc-devs/pymc#5813]
## Description of your problem

I am trying to use pm.Deterministic, by get an error related to copy(). Here is the code: 
``` python
import pymc3 as pm

def grade(i, d):
    if i == 0 and d == 0:
        cat = [0.3, 0.4, 0.3]
    elif i == 0 and d == 1:
        cat = [0.05, 0.25, 0.7]
    elif i == 1 and d == 0:
        cat = [0.9, 0.08, 0.02]
    else:
        cat = [0.5, 0.3, 0.2]
    return cat

with pm.Model() as grade_model:
    d = discrete.Bernoulli("d", 0.4, observed=0)
    i = discrete.Bernoulli("i", 0.3, observed=1)
    grade_ = pm.Deterministic("grade_",  grade(i,d))  # << generated ERROR
```
and the returned error: 
``` python
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
<ipython-input-56-53b3783b5348> in <module>
     15     d = discrete.Bernoulli("d", 0.4, observed=0)
     16     i = discrete.Bernoulli("i", 0.3, observed=1)
---> 17     grade_ = pm.Deterministic("grade_",  grade(i,d))

~/anaconda3/lib/python3.8/site-packages/pymc3/model.py in Deterministic(name, var, model, dims)
   1944     """
   1945     model = modelcontext(model)
-> 1946     var = var.copy(model.name_for(name))
   1947     model.deterministics.append(var)
   1948     model.add_random_variable(var, dims)

TypeError: copy() takes no arguments (1 given)
```
Any help on what to do is appreciated. I should note that if I use the `Deterministic` function in the following way, 
there is no error: 
``` python
pm.Deterministic("grade_", i**2), 
```
there is no error. However, `i**2` is not what I want. The output of the deterministic function `grade(i,d)` is a list. 
Thanks for any insights!

**Comment by ricardoV94** [comment]
Hello.

This is not a bug but an incorrect use of the library. Please open a topic on our [Discourse forum](https://discourse.pymc.io/) instead, as we try to keep GitHub for developer issues only.

To save you some time, PyMC variables have to be handled slightly differently than usual Python / Numpy values and arrays, and one of the things you can't do are `if` statements and variable assignments as in your helper function. This guide might help you: https://docs.pymc.io/en/v3/PyMC3_and_Theano.html

**Comment by ghost** [comment]
@erlebach You may. want to peek at the [coal mining disasters example](https://docs.pymc.io/en/v3/pymc-examples/examples/getting_started.html?highlight=coal#Case-study-2:-Coal-mining-disasters) to see how you might handle your conditional statement.

**Comment by erlebach** [comment]
Thanks, @cfonnesbeck . Here is how I solved the problem. This seems to work. However, there is an error when setting the 'l' variable to 1 in the Bernoulli. No idea why, but this seems unrelated to my original question. I posted on pymc discourse. 

``` python
import pymc3 as pm
import pymc3.distributions.discrete as discrete
import matplotlib.pyplot as plt
import arviz as az
# import aesara.tensor as at
import numpy as np
from tqdm import tqdm
import pandas as pd
from theano import tensor as tt
from theano.ifelse import ifelse

# SOME DETERMINISTIC FUNCTIONS
def grade(i, d):
    a00 = tt.and_(tt.eq(i,0), tt.eq(d,0))
    a10 = tt.and_(tt.eq(i,1), tt.eq(d,0))
    a01 = tt.and_(tt.eq(i,0), tt.eq(d,1))
    a11 = tt.and_(tt.eq(i,1), tt.eq(d,1))
    cat1 = [0.3,0.4, 0.3]
    cat2 = [0.05, 0.25, 0.7]
    cat3 = [0.9, 0.08, 0.02]
    cat4 = [0.5, 0.3,  0.2]
    return a00*cat1 + a10*cat2 + a01*cat3 + a11*cat4

def letter(g):
    return tt.eq(g, gA)*0.9 + tt.eq(g, gB)*0.6 + tt.eq(g, gC)*0.01

def score(i):
     return tt.eq(i, 0)*0.05 + tt.eq(i, 1)*0.8

gA = 2
gB = 5
gC = 10

# MODEL CREATION: observations passed in as arguments through a dictionary
def create_model(obs):
    with pm.Model() as grade_model:
        if "d" in obs.keys():
            d = discrete.Bernoulli("d", 0.4, observed=obs["d"])
        else:
            d = discrete.Bernoulli("d", 0.4)
            
        if "i" in obs.keys():
            i = discrete.Bernoulli("i", 0.3, observed=obs["i"])
        else:
            i = discrete.Bernoulli("i", 0.3)

        p_grade = grade(i, d)
        if "g" in obs.keys():
            g = discrete.Categorical("g", p_grade, observed=obs["g"])
        else:
            g = discrete.Categorical("g", p_grade)
            
        p_letter = letter(g)
        if "l" in obs.keys():
            l = discrete.Bernoulli("l", p_letter, observed=obs["l"])
        else:
            l = discrete.Bernoulli("l", p_letter)
            print("l: ", l)
            
        p_score = score(i)
        if "s" in obs.keys():
            s = discrete.Bernoulli("s", p_score, observed=obs["s"])
        else:
            s = discrete.Bernoulli("s", p_score)
    return grade_model;

# 
def run_model(model):
    with model:
        step = pm.Metropolis()
        trace = pm.sample(n_init=5000, step=step, draws=5000, tune=1000, return_inferencedata=False, progressbar=False)
        display(az.summary(trace, kind="stats"))
        lst = list(trace.points())
#         glst = [dct['g'] for dct in lst]
#         slst = [dct['s'] for dct in lst]
#         llst = [dct['l'] for dct in lst]
        return lst, trace #, [np.mean(x) for x in [glst, slst, llst]];

def run(obs):
    model = create_model(obs)
#     model = create_model_new(obs)
    lst, trace = run_model(model)
    return model, trace

def run(obs):
    model = create_model(obs)
#     model = create_model_new(obs)
    lst, trace = run_model(model)
    return model, trace

model1, trace1 = run(obs={})

# RUNS FINE. 
run(obs={'l':0.})

# PRODUCES ERROR (the observation is in a Bernoulli)
run(obs='l':1})   # <<< ERROR
```

**Comment by erlebach** [comment]
I solved my latest problem by fixing an error in the `letter` function. My error derived from the fact that I did not realize that Categorical returned a float (0., 1., 2., etc). My only question is why doesn't `Categorical` return an integer? Aren't `floats` problematic, especially when u sing them in conditions as I am doing, which is iffy at best? Thanks.
``` python
gA, gB, gC = (0., 1., 2.)
def letter(g):
    return tt.eq(g, gA)*0.9 + tt.eq(g, gB)*0.6 + tt.eq(g, gC)*0.01
```
